airflow/jobs.py

```diff
@@ -2372,20 +2389,33 @@ def _execute(self):
         executor = self.executor
         executor.start()
 
-        self._execute_for_run_dates(run_dates=run_dates,
-                                    ti_status=ti_status,
-                                    executor=executor,
-                                    pickle_id=pickle_id,
-                                    start_date=start_date,
-                                    session=session)
-
-        executor.end()
-        session.commit()
-        session.close()
+        ti_status.total_runs = len(run_dates)  # total dag runs in backfill
 
-        err = self._collect_errors(ti_status=ti_status, session=session)
-        if err:
-            raise AirflowException(err)
+        try:
+            while len(run_dates) > 0:
+                self._execute_for_run_dates(run_dates=run_dates,
+                                            ti_status=ti_status,
+                                            executor=executor,
+                                            pickle_id=pickle_id,
+                                            start_date=start_date,
+                                            session=session)
+
+                run_dates = [run_date for run_date in run_dates
```

Comment:
Sorry I meant the second run_dates assignment (the one just above this comment) could just be replaced with the any check in the if statement below. The original one can stay intact.